src/orderbook_fetcher.ts

```diff
@@ -0,0 +1,84 @@
+import Web3 from 'web3'
+import { AbiItem } from 'web3-utils'
+import { getOpenOrdersPaginated, Fraction, Orderbook, Offer, OrderBN } from '@gnosis.pm/dex-contracts'
+import { BatchExchangeViewer } from '@gnosis.pm/dex-contracts/build/types/BatchExchangeViewer'
+import batchExchangeViewerAbi from '@gnosis.pm/dex-contracts/build/contracts/BatchExchangeViewer.json'
+import { CategoryLogger } from 'typescript-logging'
+
+interface Network {
+  address: string
+}
+
+export class OrderbookFetcher {
+  orderbooks: Map<string, Orderbook> = new Map()
+  lastUpdatedBatchId = 0
+  batchExchangeViewer: BatchExchangeViewer | null = null
+
+  constructor(readonly web3: Web3, readonly pageSize: number, readonly logger: CategoryLogger) {
+    // Schedule check to update orderbook every 1s
+    setInterval(async () => {
+      await this.maybeUpdateOrderbooks()
+    }, 1000)
+  }
+
+  async loadBatchExchangeViewer(): Promise<BatchExchangeViewer> {
+    if (this.batchExchangeViewer) {
+      return this.batchExchangeViewer
+    }
+    const networkId = await this.web3.eth.getChainId()
+    if (isKeyOf(batchExchangeViewerAbi.networks, networkId)) {
+      this.batchExchangeViewer = new this.web3.eth.Contract(
+        batchExchangeViewerAbi.abi as AbiItem[],
+        (batchExchangeViewerAbi.networks[networkId] as Network).address,
+      ) as BatchExchangeViewer
+      return this.batchExchangeViewer
+    } else {
+      throw new Error(`Contract not deployed on network with ID ${networkId}`)
+    }
+  }
+
+  async maybeUpdateOrderbooks() {
+    const block = await this.web3.eth.getBlock('latest')
+    const currentBatchId = batchIdFromTimestamp(block.timestamp as number)
+    if (currentBatchId > this.lastUpdatedBatchId) {
```

Comment:
But the order book changes on every order/withdraw/deposit. Should we only update on new batches?

Reply:
No, you are right we should update periodically. This assumption came from the sovler where we are looking at the finalized orderbook which only changes on every full batch.